src/transbank.c

```diff
@@ -0,0 +1,242 @@
+#include "transbank.h"
+
+struct sp_port *port;
+
+static const int DEFAULT_TIMEOUT = 500;
+
+enum HexCodes{
+  ACK = 0x06,
+  NACK = 0x15,
+  STX = 0x02,
+  ETX = 0x03,
+  PIPE = 0x7C
+};
+
+static int BITS = 8;
+static int PARITY = SP_PARITY_NONE;
+static int STOP_BITS = 1;
+
+static char GET_TOTALS_MESSAGE[] = {0x02, 0x30, 0x37, 0x30, 0x30, 0x7C, 0x7C, 0x03, 0x04};
+static char LOAD_KEYS_MESSAGE[] = {0x02, 0x30, 0x38, 0x30, 0x30, 0x03, 0x0B};
+static char POLLING_MESSAGE[] = {0x02, 0x30, 0x31, 0x30, 0x30, 0x03, 0x02};
+static char CHANGE_TO_NORMAL_MESSAGE[] = {0x02, 0x30, 0x33, 0x30, 0x30, 0x03, 0x00};
+
+struct message_t {
+  int payloadSize;
+  int responseSize;
+  int retries;
+  char* payload;
+};
+
+static Message GET_TOTALS = {
+  .payload = GET_TOTALS_MESSAGE,
+  .payloadSize =7,
+  .responseSize = 21,
+  .retries = 1
+};
+
+static Message LOAD_KEYS = {
+  .payload = LOAD_KEYS_MESSAGE,
+  .payloadSize = 7,
+  .responseSize = 33,
+  .retries = 3
+};
+
+static Message POLLING = {
+  .payload = POLLING_MESSAGE,
+  .payloadSize = 7,
+  .responseSize = 1,
+  .retries = 3
+};
+
+static Message CHANGE_TO_NORMAL = {
+  .payload = CHANGE_TO_NORMAL_MESSAGE,
+  .payloadSize = 7,
+  .responseSize = 1,
+  .retries = 3
+};
+
+void print_ports(){
+  struct sp_port **ports;
+  int retval = sp_list_ports(&ports);
+  for (int i=0; ports[i] != NULL; i++){
+    printf("%s\n", sp_get_port_name(ports[i]));
+  }
+}
+
+char* list_ports() {
+  struct sp_port **ports;
+  char* portList;
+
+  int retval = sp_list_ports(&ports);
+
+  if (retval == SP_OK) {
+    int separators = 0;
+    int chars = 0;
+    for (int i=0; ports[i] != NULL; i++){
```

Comment:
⚠️ This is a Question and maybe a suggestion, but you can ignore it:

I saw that you're using libserialport. Therefore, sp_list_ports fills `ports` with an array of pointers to sp_port structures, terminated by a NULL. 

In that case, `sizeof(ports)/sizeof(ports[0]);` could return a possible value for separators? (I don't know if the NULL element is malloc'ed, affecting the `sizeof(ports)` result.

Reply:
It works if the structure size is fixed, but serial-port names are string with different sizes.